lib/evaluate/fold.cc

```diff
@@ -392,10 +392,9 @@ Expr<Type<TypeCategory::Integer, KIND>> FoldOperation(FoldingContext &context,
                 std::is_same_v<From, Expr<SomeInteger>> ||
                 std::is_same_v<From, Expr<SomeComplex>>) {
               return Fold(context, ConvertToType<T>(std::move(x)));
-            } else {
-              common::die("int() argument type not valid");
-              return Expr<T>{std::move(funcRef)};  // unreachable
             }
+            common::die("int() argument type not valid");
+            return Expr<T>{std::move(funcRef)};  // unreachable
```

Comment:
Does `return {};` fail to compile here?  I see the unreachable comment, but there is still a fair number of characters that look like they could do something before that.

Reply:
Yeah,` return {};` would require a constructor without arguments and Expr<T> do not have such constructor on purpose (they use EVALUATE_UNION_CLASS_BOILERPLATE from lib/evaluate/common.h that explicitly delete the constructor without arguments).
The code should compile without the return anyway because `common::die` is `[no return]` but we would get warnings from gcc....

Reply:
I thought so.  What about wrapping that dummy `Expr<T>` ctor in a function and call it something like `UnreachableFakeValue()`?  It's fine with me to leave it as well.

Reply:
That is not a bad idea, I will not make it in this PR but maybe we could have in one of our common header something like:

```
template<typename T>
T UnreachableFakeValue() {
  common::die("Reached statement that is intended to be unreachable");
  return *reinterpret_cast<T*>(nullptr);
}
```
 We could use it here with:
`return UnreachableFakeValue<Expr<T>>();`

It would make the intention clear and offer dynamic checking.

Reply:
Anyway, I may be wrong and the [no return] aspect of common::die is sufficient here; gcc 7.2, gcc 8.1 and clang-7 are not giving warning if I completely remove the return. I think I have added this first for the lambda return type deduction but it is not useful anymore because I am specifying the return type explicitly. I will re-check with more compiler and remove the return later if I do not get warnings.